Improve usability of code base (#38)
m2lines/CAM-ML: README.md

# CAM-ML: A Fork of The Community Atmosphere Model Implementing a Machine Learning Convection Parameterisation

> This Fork of the CAM model implements a new convection parameterisation, YOG.

Reviewer: Perhaps link to the convection-parameterisation-in-cam repository?

Author: Added that, but am not sure whether it's weird to link to another repository in the first sentence of the README -- might not be clear what the relationship between the repositories is

  _Use of Neural Networks for Stable, Accurate and Physically Consistent Parameterization of Subgrid Atmospheric Processes With Good Performance at Reduced Precision_
  Geophysical Research Letters, 48, e2020GL091363 (2021). DOI: [10.1029/2020GL091363](https://doi.org/10.1029/2020GL091363)

The work is contained in a `CAM-ML` branch which is based off the `cam_cesm2_1_rel_60` tag.

> Note that the `doc` folder and `README_EXTERNALS` are not updated, but just artifacts from the fork that we are keeping to be able to merge back to CESM later. 

Reviewer: ```suggestion
Note that the `doc` folder and `README_EXTERNALS` are not updated, but just artifacts from the fork that we are keeping to be able to merge back to CESM later.
Information specific to running this fork of CAM within CESM is included in this README.
```

## Using this model in a CESM Run

> This section describes how to use CAM-ML as an external plug-in in a normal CESM run. 

Reviewer: ```suggestion
This section describes how to use CAM-ML as the atmospheric component in a normal CESM run. 
```

```
./manage_externals/checkout_externals
```

> to fetch the external components. 

Reviewer: Not sure why it has added a whitespace to the end of this line.

Author: tried to remove it

```
from `<cesm_root>/cime/scripts/`.

The `<testcase_directory>` should be a separate directory outside of the code directory, to avoid cluttering up the local repository.

Once this has been done then edit `user_nl_cam` for the case as detailed below.
This is a CAM namelist generated from the default for the case.
Add the following lines:

> 1. `deep_scheme = 'off'`\

Reviewer: I'd add `yog_scheme = 'on'` as a separate line (step 2?), as without it you will be running with no convection scheme at all.

For explanation line: we turn deep_scheme off as YOG covers the deep convection processes.
We have to turn yog on as there may be other cases where someone does not want to use it, and it should not be turned on by default.

Author: Changed that

    If you want to run with the new YOG convection scheme, add `yog_scheme = 'on'`, as 	well.\
    If running a comparison to the ZM scheme also add `run_deep_comp = 'on'`.
2. `yog_nn_weights = '<PATH/TO/WEIGHTS.nc>'`\
    The path to the NN weights. There are some weights in `src/physics/cam/' of this respository (CAM-ML) which can be used, or they can be generated from the [standalone model](https://github.com/m2lines/convection-parameterization-in-CAM).
3. `SAM_sounding = '<PATH/TO/SAM/SOUNDING.nc>'`\
    The path to the SAM sounding for the NN.\
    This file is generated using the `sounding_to_netcdf.py` script in the resources of the [standalone NN code](https://github.com/m2lines/convection-parameterization-in-CAM), and it can be just copied over to a suitable place.

All the paths have to be absolute, as they will be used when the code is run on the compute nodes.

We can then run `./case.setup` and `./case.build`to setup and build the test case, and `./case.submit` to submit the job to the scheduler.

**Note:**  
By default, CESM will place output on NCAR's supercomputer Derecho in `/glade/scratch/user/case/`
and logs/restart files in `/glade/scratch/user/archive/case/`.

> To place all output with logs in `archive/case` switch 'short term archiving' on by editing `env_run.xml` in the case directory to change `DOUT_S` from `FALSE` to `TRUE`.

Reviewer: I think the inverse is now true - successful results get archived by default.

Failed jobs are left in `/glade/scratch/user/case/`.

If running short-term or test runs it makes sense not to archive (extra time, job, and data).

Rather than editing `env_run.xml` I would use the `﻿﻿./xmlchange DOUT_S=FALSE` functionality from the case directory.

Author: Added that

We can then run `./case.setup` and `./case.build`to setup and build the test case, and `./case.submit` to submit the job to the scheduler.

**Note:**  
By default, CESM will place outputs and logs/restart files on Derecho in `/glade/derecho/scratch/<user>/archive/<case>/`.
Outputs of failed runs can be found in`/glade/derecho/scratch/<user>/<case>/`.

To place all output with logs in `archive/case` switch 'short term archiving' on by editing `env_run.xml` in the case directory to change `DOUT_S` from `FALSE` to `TRUE` -- you can do this quickly by running `﻿./xmlchange DOUT_S=FALSE`.

### Expected Output

> The run will create a `bld` and a `run` directory in the case folder. The `bld` directory contains build files, logs and executables. The `run` directory contains the run logs, namelists and output Netcdf files.

Reviewer: Not allowing me to make a suggestion for some reason, but "NetCDF".

Author: Is there no "role eyes" emoji on here....?

Author: done

> 

Reviewer: Specify that they will be in the OUTPUT case folder IN SCRATCH?
As opposed to the case folder where you are running the submission etc from.

Author: done

A successful model run will have the line \
`******* END OF MODEL RUN *******` \

> at the bottom of the run log file (e.g., for the atmosphere run, `atm.log.<stuff>.gz`).

Reviewer: I would specify that we look for this at the end of the atmosphere logfile.

The cesm log file doesn't really indicate it has reached the end successfully or not, and I'm not really sure what appears in the ocean/other logfiles at the moment.

Author: done

CAM6 namelist settings - http://www.cesm.ucar.edu/models/cesm2/settings/current/cam_nml.html

Please see the [wiki](https://github.com/ESCOMP/CAM/wiki) for information.

## Contributing

Contributions to the repository are welcome.

Open tickets can be viewed at [**Issues**](https://github.com/m2lines/CAM-ML/issues).

> To contribute, find a relevant issue or open a new one and assign yourself to work on it. Then create a branch in which to add your contribution and open a pull request. Once ready assign a reviewer and request a code review. Merging should only be performed once a code review has taken place.

Reviewer: I'm not sure this is entirely true.
Would it be better to suggest people comment on an issue if they think that they can tackle an issue and would like to contribute?
We are unlikely to take contributions at the moment from outside ICCS/M2LInES, or indeed from people outside this project for now, particularly without discussion beforehand given the need to understand what we are doing in this project.

Author: adapted the text accordingly
